File: IKVM.OpenJDK/src/classpath/ikvm/internal/io/FileStreamExtensions.java
package ikvm.internal.io;

import cli.System.IO.FileAccess;
import cli.System.IO.FileMode;
import cli.System.IO.FileOptions;
import cli.System.IO.FileShare;
import cli.System.IO.FileStream;
import cli.System.Security.AccessControl.FileSystemRights;

import ikvm.lang.Internal;

@Internal
public final class FileStreamExtensions
{

    public static FileStream create(String path, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options) throws java.io.IOException
    {
        return createInternal(path, mode, rights, share, bufferSize, options);
    }

    static native FileStream createInternal(String path, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options) throws java.io.IOException;

}

File: src/IKVM.Runtime/Java/Externs/ikvm/internal/io/FileStreamExtensions.cs
using System.IO;
using System.Security.AccessControl;

namespace IKVM.Java.Externs.ikvm.@internal.io
{

    static class FileStreamExtensions
    {

        public static FileStream createInternal(string path, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options)
        {
#if NETFRAMEWORK
            return new FileStream(path, mode, rights, share, bufferSize, options);
#else
            var access = 0;
            if ((rights & FileSystemRights.Read) != 0)
                access |= (int)FileAccess.Read;
            if ((rights & (FileSystemRights.Write | FileSystemRights.AppendData)) != 0)
                access |= (int)FileAccess.Write;
            if (access == 0)
                access = (int)FileAccess.ReadWrite;

            return new FileStream(path, mode, (FileAccess)access, share, bufferSize, options);
#endif
        }

    }

}
